lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -344,6 +344,12 @@ bool SILCombiner::tryOptimizeKeypathOffsetOf(ApplyInst *AI,
   KeyPathPattern *pattern = kp->getPattern();
   SubstitutionMap patternSubs = kp->getSubstitutions();
   CanType rootTy = pattern->getRootType().subst(patternSubs)->getCanonicalType();
+
+  // TODO: support lowering of the rootTy if it's not a legal SIL type in the
+  // first place, e.g. if it contains an AnyFunctionType.
+  if (!rootTy->isLegalSILType())
+    return false;
```

Comment:
Instead of skipping out on these cases, it's not that hard to handle them correctly. The lowered type of the ends of a keypath projection are always going to be the maximally-abstracted form of the formal type, so you can do

```
CanType rootTy = AI->getFunction()->getLoweredType(AbstractionPattern::getOpaque(),
    pattern->getRootType().subst(patternSubs)->getCanonicalType());
```

And that should work for all root types. Even if the formal type happens to be a valid lowered SIL type, it is still a category error to use a formal type without lowering, and there could remain problems from doing so.

Reply:
ok, thanks!
-> https://github.com/apple/swift/pull/62862